test/boost/row_cache_test.cc

```diff
@@ -4219,3 +4219,58 @@ SEASTAR_TEST_CASE(test_reading_of_nonfull_keys) {
                     .produces(m1);
         });
 }
+
+SEASTAR_TEST_CASE(test_populating_cache_with_expired_and_nonexpired_tombstones) {
+        return do_with_cql_env_thread([](cql_test_env& env) {
+            replica::database& db = env.local_db();
+            auto& mm = env.migration_manager().local();
+            tests::reader_concurrency_semaphore_wrapper semaphore;
+
+            sstring ks_name = "ks";
+            sstring table_name = "table_name";
+
+            schema_ptr s = schema_builder(ks_name, table_name)
+                    .with_column("pk", int32_type, column_kind::partition_key)
+                    .with_column("ck", int32_type, column_kind::clustering_key)
+                    .build();
+            mm.announce(
+                mm.prepare_new_column_family_announcement(s, api::new_timestamp()).get(),
+                mm.start_group0_operation().get()
+            ).get();
+
+            replica::table& t = db.find_column_family(ks_name, table_name);
+
+            auto pkey = dht::decorate_key(*s, partition_key::from_single_value(*s, serialized(1)));
+            auto ck1 = clustering_key::from_deeply_exploded(*s, {1});
+            auto ck1_prefix = clustering_key_prefix::from_deeply_exploded(*s, {1});
+            auto ck2 = clustering_key::from_deeply_exploded(*s, {2});
+            auto ck2_prefix = clustering_key_prefix::from_deeply_exploded(*s, {2});
+
+            auto dt_noexp = gc_clock::now();
+            auto dt_exp = gc_clock::now() - std::chrono::seconds(s->gc_grace_seconds().count() + 1);
+
+            mutation m(s, pkey);
+            m.partition().apply_delete(*s, ck1_prefix, tombstone(1, dt_noexp)); // create non-expired tombstone
+            m.partition().apply_delete(*s, ck2_prefix, tombstone(2, dt_exp)); // create expired tombstone
+            t.apply(m);
+            t.flush().get();
+
+            // Clear the cache and repopulate it by reading sstables
+            t.get_row_cache().evict();
+
+            auto reader = t.get_row_cache().make_reader(s, semaphore.make_permit());
+            deferred_close dc{reader};
+            reader.consume_pausable([s](mutation_fragment_v2&& mf) {
+                return stop_iteration::no;
+            }).get();
+
+            cache_entry& entry = t.get_row_cache().lookup(pkey);
+            auto& cp = entry.partition().version()->partition();
+
+            BOOST_REQUIRE_EQUAL(cp.tombstone_for_row(*s, ck1), row_tombstone(tombstone(1, dt_noexp))); // non-expired tombstone is in cache
+            BOOST_ASSERT(cp.find_row(*s, ck2) == nullptr); // expired tombstone isn't in cache
```

Comment:
Better to use BOOST_REQUIRE, to avoid crashing if the assert fails (and to prevent dependency on NDEBUG).

Does the test fail without the patch?

Reply:
Thanks, changed to BOOST_REQUIRE

All tests passed locally (tried few times)
I'd appreciate if you told me name of the test failed: I have no access to see tests details

Reply:
No tests fail for me (didn't try). I'm asking, if you run the updated row_cache_test.cc without the updated row_cache.cc, will the test fail? To prove there isn't some mistake in the test that would cause it to pass either way.

Reply:
yes, when I run row_cache_test.cc with new test without changes in code it fails, I've checked it.
There both tombstones get in the row cache after repopulating - seems changes made in table.cc (we didn't update row_cache.cc) works as expected